core/src/main/java/org/elasticsearch/action/support/single/shard/TransportSingleShardAction.java

```diff
@@ -173,7 +173,7 @@ public void handleResponse(final Response response) {
 
                     @Override
                     public void handleException(TransportException exp) {
-                        perform(exp);
+                        listener.onFailure(exp);
```

Comment:
If you do that then the action will not be retried in another shard. It's ok for the analyze api because it would fail on every shard but it seems like it would break other actions. I think the problem you're trying to fix is that perform does not check if shardIt is null before accessing it.